src/js/trove/matrix.js

```diff
@@ -0,0 +1,1299 @@
+                }
+                ans_vectors[col] = currVector
+                magSqr(col);
+            }
+            mags = mags.map(v => runtime.num_sqrt(v));
+            var new_arr = new Array(self.$h * self.$w);
+            for (i = 0; i < self.$h; i++) {
+                for (j = 0; j < self.$w; j++) {
+                    new_arr[get1dpos(i, j, self.$w)] = runtime.divide(ans_vectors[j].$underlyingMat[i], mags[j]);
+                }
+            }
+
+            return makeMatrix(self.$h, self.$w, new_arr);
+
+
+
+        }
+        //Finds QR decomposition using Gram Schmidt
+        var qrMatrix = function (self) {
+            arity(1, arguments, "qr-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "qr-mat", self, annMatrix);
+            q = Gram_Schmidt(self);
+            r = multMatrix(transposeMatrix(q), self);
+            return runtime.makeTuple([q, r]);
+        }
+        //Finds pth norm of matrix
+        var normMatrix = function (self, p) {
+            arity(2, arguments, "norm-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "norm-mat", self, annMatrix, p, runtime.Number);
+            sum = self.$underlyingMat.reduce((a, v) => runtime.plus(a, runtime.num_expt(v, p)), 0);
+            return runtime.num_expt(sum, runtime.divide(1, p));
+        }
+        //Creates matrix of size (h,w) by applying f
+        var buildMatrix = function (h, w, f) {
+            arity(3, arguments, "build-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "build-mat", h, runtime.Number, w, runtime.Number, f, runtime.Function);
+            posInteger(h, w);
+            return runtime.safeCall(
+                function () {
+                    return runtime.raw_array_build(F(function (idx) {
+                        res = f.app((idx - (idx % w)) / w, idx % w);
+                        if (!jsnum.isPyretNumber(res)) {
+                            return runtime.ffi.throwMessageException("Function must return a number");
+                        }
+                        return res;
+                    }), h * w);
+                }, function (result) {
+                    return makeMatrix(h, w, result);
+                }
+            );
+        }
+        //Calculates forbenius Matrix
+        var frobMatrix = function (self) {
+            arity(1, arguments, "frob-norm", false);
+            runtime.checkArgsInternalInline("Matrix", "frob-norm", self, annMatrix);
+            return runtime.num_sqrt(traceMatrix(multMatrix(transposeMatrix(self), self)));
+
+        }
+        //Calculates A^n 
+        var expMatrix = function (self, n) {
+            arity(2, arguments, "exp-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "exp-mat", self, annMatrix, n, runtime.Number);
+            posInteger(n)
+            res = duplicateMatrix(self);
+            for (i = 1; i < n; i++) {
+                res = multMatrix(res, self);
```

Comment:
If `multMatrix` uses safeCalls, then _callers_ of `multMatrix` must use it in `safeCalls`: you need to now wrap this in a `raw_array_fold` to do the looping the right number of times -- you can use a bogus array of e.g. `new Array(n)`, because you don't actually care about the contents of the array; you just care about the matrix-multiplication result.

(And if you want to get more efficient here, you should use [exponentiation by squaring](https://en.wikipedia.org/wiki/Exponentiation_by_squaring) to reduce the number of matrix-multiplications you have to do.

Reply:
Is this for every single call of `multMatrix`. That is in some function if I cal`l c = multMatrix(b,c)` does that need to be wrapped in `safeCall` now?